pkg/tfbridge/provider.go

```diff
@@ -59,6 +59,96 @@ type Resource struct {
 	TFName string           // the Terraform resource name.
 }
 
+// runTerraformImporter runs the Terraform Importer defined on the Resource for the given
+// resource ID, and returns a replacement input map if any resources are matched.
+func (res *Resource) runTerraformImporter(resourceID resource.ID) (map[string]string, error) {
+	// There is nothing to do here if the resource doesn't have an importer defined in the
+	// Terraform schema.
+	if res.TF.Importer == nil {
+		return nil, nil
+	}
+
+	glog.V(9).Infof("%s has TF Importer", res.TFName)
+
+	id := resourceID.String()
+
+	// Prepare a Terraform ResourceData for the importer
+	data := res.TF.Data(nil)
+	data.SetId(id)
+	data.SetType(res.TFName)
+
+	// Run the importer defined in the Terraform resource schema
+	results, err := res.TF.Importer.State(data, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "importing %s", id)
+	}
+
+	// No resources were returned - error out
+	if len(results) < 1 {
+		return nil, errors.Errorf("importer for %s returned no resources", id)
+	}
+
+	// Allow constructing an error in the case that we have a nil InstanceState returned from
+	// Terraform, which is always a programming error.
+	makeNilStateError := func(badResourceID string) error {
+		return errors.Errorf("importer for %s returned a empty resource state. This is always "+
+			"the result of a bug in the resource provider - please report this "+
+			"as a bug in the Pulumi provider repository.", badResourceID)
+	}
+
+	// A Terraform importer can return multiple ResourceData instances for different resources. For
+	// example, an AWS security group will also import the related security group rules as independent
+	// resources.
+	//
+	// Some Terraform importers _change_ the ID of the resource to allow for multiple formats to be
+	// specified by a user (for example, an AWS API Gateway Response). In the case that we only have
+	// a single ResourceData returned, we will use that ResourceData regardless of whether the ID
+	// matches, provided the resource Type does match.
+	//
+	// If we get multiple ResourceData back, we need to search the results for one which matches both
+	// the Type and ID of the resource we were trying to import (the "primary" InstanceState).
+	//
+	// The Type can be identified by looking at the ephemeral data attached to the InstanceState, since
+	// it is not stored in all cases - only for import.
+	var primaryInstanceState *terraform.InstanceState
+
+	if len(results) == 1 {
+		// Take the only result, assuming the Type matches
+		state := results[0].State()
+		if state == nil {
+			return nil, makeNilStateError(id)
+		}
+		if state.Ephemeral.Type == res.TFName {
+			primaryInstanceState = state
+		}
+	} else {
+		// Search for a Type+ID match, and use the first (if any)
+		for _, result := range results {
+			if result.Id() != id {
```

Comment:
I thought I remembered cases where we've seen ID mismatches between the input ID and the imported ID. Should we call out that possibility here?

Reply:
We decided the last time around that if this was the case and only one resource was returned, we would use that resource, but otherwise we considered this irreconcilable.

Reply:
Can we leave a comment to that effect?

Reply:
I've added a much longer form comment around the behaviour and rationale above this block.